File: PokeScan/Services/SetResolver.swift
import Foundation

final class SetResolver {
    private let entries: [SetEntry]

    init(entries: [SetEntry]) {
        self.entries = entries
    }

    /// Resolves setCode from "NNN/TTT" format and card language.
    /// Returns "unknown" if resolution fails.
    func resolve(setNumber: String, language: CardLanguage) -> String {
        guard let total = parseTotal(from: setNumber),
              let cardNum = parseCardNumber(from: setNumber)
        else { return "unknown" }

        let langStr = language.rawValue
        let candidates = entries.filter { $0.total == total && $0.language == langStr }

        guard !candidates.isEmpty else { return "unknown" }
        if candidates.count == 1 { return candidates[0].setCode }

        // When printedTotal is available, use it to disambiguate sets with same total.
        // Example: base1 (printedTotal=102) vs ex5/Hidden Legends (printedTotal=101).
        let withPrinted = candidates.filter { $0.printedTotal != nil }
        if withPrinted.count == candidates.count {
            let byPrinted = withPrinted.filter { $0.printedTotal == total }
            if byPrinted.count == 1 { return byPrinted[0].setCode }
        }

        // Fallback: newest set wins. Ties broken by setCode ascending for determinism.
        let valid = candidates
            .filter { cardNum <= $0.total }
            .sorted {
                if $0.releaseYear != $1.releaseYear { return $0.releaseYear > $1.releaseYear }
                return $0.setCode < $1.setCode
            }
        return valid.first?.setCode ?? "unknown"
    }

    private func parseTotal(from setNumber: String) -> Int? {
        let parts = setNumber.split(separator: "/")
        guard parts.count == 2, let total = Int(parts[1]) else { return nil }
        return total
    }

    private func parseCardNumber(from setNumber: String) -> Int? {
        let parts = setNumber.split(separator: "/")
        guard parts.count == 2, let num = Int(parts[0]) else { return nil }
        return num
    }
}

File: android/app/src/main/java/com/pokescan/app/data/service/SetResolver.kt
package com.snapdex.app.data.service

import com.snapdex.app.domain.model.CardLanguage
import com.snapdex.app.domain.model.SetEntry
import javax.inject.Inject

data class ResolvedSet(
    val setCode: String,
    val setName: String?,
    val releaseYear: Int?,
    val candidates: List<SetEntry>? = null,
)

class SetResolver @Inject constructor() {

    fun resolve(entries: List<SetEntry>, setNumber: String, language: CardLanguage): ResolvedSet {
        val parts = setNumber.split("/")
        if (parts.size != 2) return ResolvedSet("unknown", null, null)
        val cardNum = parts[0].toIntOrNull() ?: return ResolvedSet("unknown", null, null)
        val total = parts[1].toIntOrNull() ?: return ResolvedSet("unknown", null, null)

        val candidates = entries.filter { it.total == total && it.language == language.raw }
        if (candidates.isEmpty()) return ResolvedSet("unknown", null, null)
        if (candidates.size == 1) {
            val e = candidates[0]
            return if (cardNum in 1..(total * 2)) ResolvedSet(e.setCode, e.name, e.releaseYear) else ResolvedSet("unknown", null, null)
        }

        // printedTotal disambiguation — base1 (printedTotal=102) vs ex5 (printedTotal=101)
        // Only applies when ALL candidates have printedTotal, to avoid partial data skewing result
        val withPrinted = candidates.filter { it.printedTotal != null }
        if (withPrinted.size == candidates.size) {
            val byPrinted = withPrinted.filter { it.printedTotal == total }
            if (byPrinted.size == 1) {
                val e = byPrinted[0]
                return ResolvedSet(e.setCode, e.name, e.releaseYear)
            }
        }

        // Newest-wins fallback: deterministic — releaseYear DESC, then setCode ASC
        // candidates field populated so callers can attempt pHash disambiguation
        val eligible = candidates.filter { cardNum <= it.total }
        val winner = eligible
            .sortedWith(compareByDescending<SetEntry> { it.releaseYear }.thenBy { it.setCode })
            .firstOrNull()
        return if (winner != null) {
            ResolvedSet(winner.setCode, winner.name, winner.releaseYear, candidates = eligible)
        } else {
            ResolvedSet("unknown", null, null)
        }
    }
}
